🚀  Like Button

paOmer opened the pull request:
Setting up the like button.
- Added the like button functionality in the `feed/views.py` file.
- Set a like URL which include the post to like ID
- Links the existing dummy like button to the url
- Added tests

Only an authenticated user can like a post.

Future Improvements:
- Prevent the page from completely reload when hitting the like button. (#112)
- If an authenticated user have liked a post, he should see the like button as a filled thumbs up pic instead of the empty existing one (#113)

Close #110

ShirleyFichman reviewed (approved):
LGTM

DeanBiton reviewed (approved):
Looking good
👍

Yarboa commented on feed/views.py:
```diff
@@ -13,3 +15,17 @@ def feed(request):
+    else:
+        messages.warning(request, "You must login in order to like a post")
+
+    return HttpResponseRedirect(request.META.get('HTTP_REFERER'))
```
Why not 
return redirect('login') ?

paOmer replied:
done

Yarboa commented on feed/tests/test_feed_views.py:
```diff
@@ -45,3 +52,60 @@ def test_detail_view_page_for_invalid_post_id(self, client):
+        # Testing that when logged out, if trying to access the like url
+        # a warning message returns
+        post_to_check_likes = Post.posts.filter(title=post.title, content=post.content, author=post.author).first()
+        response = client.get(post_like_url(post_to_check_likes.id))
```
I would advisde. to check that if user is not logged it, he is forwarded to login page.

paOmer replied:
added

Yarboa reviewed (commented):
the way authenticate is used here should be revised
https://docs.djangoproject.com/en/4.0/topics/auth/default/#authenticating-users
It is a low level authentication, 
Dont you think the logic should be done ins the App side?

paOmer commented:
> the way authenticate is used here should be revised https://docs.djangoproject.com/en/4.0/topics/auth/default/#authenticating-users It is a low level authentication, Dont you think the logic should be done ins the App side?

Updated, you could find the updates in the test file added to this PR in lines 139 and 157.
Hope that I understood correctly.

Yarboa commented on feed/tests/test_feed_views.py:
```diff
@@ -131,3 +144,62 @@ def test_redirect_after_successful_post_creation(self, users, client):
+            author=liked_post.author).first()
+        like_user = authenticate(username=USERNAME1, password=USER_PASS)
+
+        if like_user is not None:
```
same as above

paOmer replied:
Done

Yarboa reviewed (changes requested):
Please see this as sample tests
https://github.com/redhat-beyond/beyond-tutorial/blob/main/msgboard/tests.py#L43

paOmer commented on feed/tests/test_feed_views.py:
```diff
@@ -131,3 +144,62 @@ def test_redirect_after_successful_post_creation(self, users, client):
+    def test_like_url_adds_like_to_post(self, post, client):
+        # Testing that the like url, adds a like to a specific post
+        post_to_like = Post.posts.filter(title=post.title, content=post.content, author=post.author).first()
+        like_user = authenticate(username=USERNAME1, password=USER_PASS)
```
Regarding the use of authenticate, I thought that what you've meant [here](https://github.com/redhat-beyond/JobSeeker/pull/111#pullrequestreview-950937959), but I've updated the tests to use the `logged_in_client` fixture which uses `client.force_login`.
Also, split to two test classes, one for logged users and the other for non logged in users.

paOmer replied:
Regrading the fixture, I've updated the tests to make a direct use of the `post` fixture instead of filtering through the DB.

Yarboa commented on feed/tests/test_feed_views.py:
```diff
@@ -159,47 +165,32 @@ def test_like_view_return_warning_when_logged_out(self, post, client):
+    def test_like_url_adds_like_to_post(self, post, logged_in_client):
+        # Testing that the like url, adds a like to a specific post
+        post_to_like = post
+        assert post_to_like.likes.count() == 0
```
test verification should be done through client url, isnt it?

paOmer replied:
Done

Yarboa commented on feed/tests/test_feed_views.py:
```diff
@@ -159,47 +165,32 @@ def test_like_view_return_warning_when_logged_out(self, post, client):
+        post_to_like = post
+        assert post_to_like.likes.count() == 0
+        logged_in_client.get(post_like_url(post_to_like.id))
+        assert post_to_like.likes.count() == 1
```
same as above those are client tests

paOmer replied:
Done

Yarboa commented on feed/tests/test_feed_views.py:
```diff
@@ -159,47 +165,32 @@ def test_like_view_return_warning_when_logged_out(self, post, client):
-        else:
-            # Login details are invalid
-            assert False
+        post_to_unlike = liked_post
```
If the test, tests the like response 
All the checks here need to read the content from response, 

response = logged_in_client.get(post_like_url(post_to_unlike.id))

You should assert content of the response generated by view

paOmer replied:
Done

paOmer commented:
Hi @Yarboa,
Changed the tests to assert content of the response context.
Also added another test `test_after_like_redirect_to_detail_view` to check redirection after success like when HTML REFFER is disabled, but couldn't figure out how or if to check redirection when HTML REFFER is allowed.

Yarboa commented:
> Hi @Yarboa, Changed the tests to assert content of the response context. Also added another test `test_after_like_redirect_to_detail_view` to check redirection after success like when HTML REFFER is disabled, but couldn't figure out how or if to check redirection when HTML REFFER is allowed.

You should run another request with client to the redirect URL
you should reach the redirected page, with 200OK and context

paOmer commented:
> > Hi @Yarboa, Changed the tests to assert content of the response context. Also added another test `test_after_like_redirect_to_detail_view` to check redirection after success like when HTML REFFER is disabled, but couldn't figure out how or if to check redirection when HTML REFFER is allowed.
> 
> You should run another request with client to the redirect URL you should reach the redirected page, with 200OK and context

I've added the 200 status code in the `test_like_url_unlikes` test.
And I also have this test:
```py
def test_after_like_redirect_to_detail_view(self, post, logged_in_client):
        # Testing that after successful like, if the user got to the like URL
        # by typing it, he would get redirected to the post detail view page.
        # Also, if a user turned off the HTTP_REFERER option in his browser he would
        # get returned to the post detail view
        response = logged_in_client.get(post_like_url(post.id))
        assert response.status_code == REDIRECT_URL_STATUS
        assert response.url == POST_DETAIL_URL + f'{post.id}/'
```

But I think there is another test missing and I cant figure out how to set it, I tried:
```py
def test(self, post, logged_in_client):
        logged_in_client.get(FEED_URL)
        response = logged_in_client.get(post_like_url(post.id))
        assert response.url == FEED_URL
```

but I dont get the feed to be the response.url, I get the post detail view as the test before.
What im trying to test here are lines 57-59 in the `feed/views.py`:
```py
 origin_url = request.META.get('HTTP_REFERER')
        if origin_url is not None:
            return HttpResponseRedirect(origin_url)
```

Yarboa commented:
> > > Hi @Yarboa, Changed the tests to assert content of the response context. Also added another test `test_after_like_redirect_to_detail_view` to check redirection after success like when HTML REFFER is disabled, but couldn't figure out how or if to check redirection when HTML REFFER is allowed.
> > 
> > 
> > You should run another request with client to the redirect URL you should reach the redirected page, with 200OK and context
> 
> I've added the 200 status code in the `test_like_url_unlikes` test. And I also have this test:
> 
> ```python
> def test_after_like_redirect_to_detail_view(self, post, logged_in_client):
>         # Testing that after successful like, if the user got to the like URL
>         # by typing it, he would get redirected to the post detail view page.
>         # Also, if a user turned off the HTTP_REFERER option in his browser he would
>         # get returned to the post detail view
>         response = logged_in_client.get(post_like_url(post.id))
>         assert response.status_code == REDIRECT_URL_STATUS
>         assert response.url == POST_DETAIL_URL + f'{post.id}/'
> ```
> 
> But I think there is another test missing and I cant figure out how to set it, I tried:
> 
> ```python
> def test(self, post, logged_in_client):
>         logged_in_client.get(FEED_URL)
>         response = logged_in_client.get(post_like_url(post.id))
>         assert response.url == FEED_URL
> ```
> 
> but I dont get the feed to be the response.url, I get the post detail view as the test before. What im trying to test here are lines 57-59 in the `feed/views.py`:
> 
> ```python
>  origin_url = request.META.get('HTTP_REFERER')
>  if origin_url is not None:
>         return HttpResponseRedirect(origin_url)
> ```

Just review the UI talk, you have those sample tests.
Once you received redirect url, you should call from the client again


>    response = client.get('/posts')
>    assert response.status_code == 301
>    response = client.get(response.url)
>    assert response.status_code == 200 
>    assert isinstance(response.context['blog_list'], list)

amittcohen reviewed (approved):
LGTM

paOmer commented:
> > > > Hi @Yarboa, Changed the tests to assert content of the response context. Also added another test `test_after_like_redirect_to_detail_view` to check redirection after success like when HTML REFFER is disabled, but couldn't figure out how or if to check redirection when HTML REFFER is allowed.
> > > 
> > > 
> > > You should run another request with client to the redirect URL you should reach the redirected page, with 200OK and context
> > 
> > 
> > I've added the 200 status code in the `test_like_url_unlikes` test. And I also have this test:
> > ```python
> > def test_after_like_redirect_to_detail_view(self, post, logged_in_client):
> >         # Testing that after successful like, if the user got to the like URL
> >         # by typing it, he would get redirected to the post detail view page.
> >         # Also, if a user turned off the HTTP_REFERER option in his browser he would
> >         # get returned to the post detail view
> >         response = logged_in_client.get(post_like_url(post.id))
> >         assert response.status_code == REDIRECT_URL_STATUS
> >         assert response.url == POST_DETAIL_URL + f'{post.id}/'
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > But I think there is another test missing and I cant figure out how to set it, I tried:
> > ```python
> > def test(self, post, logged_in_client):
> >         logged_in_client.get(FEED_URL)
> >         response = logged_in_client.get(post_like_url(post.id))
> >         assert response.url == FEED_URL
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > but I dont get the feed to be the response.url, I get the post detail view as the test before. What im trying to test here are lines 57-59 in the `feed/views.py`:
> > ```python
> >  origin_url = request.META.get('HTTP_REFERER')
> >  if origin_url is not None:
> >         return HttpResponseRedirect(origin_url)
> > ```
> 
> Just review the UI talk, you have those sample tests. Once you received redirect url, you should call from the client again
> 
> > response = client.get('/posts')
> > assert response.status_code == 301
> > response = client.get(response.url)
> > assert response.status_code == 200
> > assert isinstance(response.context['blog_list'], list)

Thanks done, also moved the like count to a fixture.

Yarboa commented on feed/tests/test_feed_views.py:
```diff
@@ -131,3 +160,60 @@ def test_redirect_after_successful_post_creation(self, users, client):
+        # a warning message returns
+        post_to_like = post
+        response = client.get(post_like_url(post_to_like.id))
+        messages = [m.message for m in get_messages(response.wsgi_request)]
```
Did you try this ?
esponse.context['messages'] instead?

paOmer replied:
I did but with no success trying to get the message content or details.
What I get from `response.context['messages']` is `<FallbackStorage: request=<WSGIRequest: GET '/login/'>>`
